File: EXP_5/bfs.c
#include <bits/stdc++.h>
using namespace std;
typedef pair<int, int> pi;
vector<vector<pi> > graph;

void addedge(int x, int y, int cost)
{
	graph[x].push_back(make_pair(cost, y));
	graph[y].push_back(make_pair(cost, x));
}

void best_first_search(int source, int target, int n)
{
	vector<bool> visited(n, false);
	priority_queue<pi, vector<pi>, greater<pi> > pq;
	pq.push(make_pair(0, source));
    visited[0]=true;
	while (!pq.empty()) {
		int x = pq.top().second;
	
		cout << x << " ";
		pq.pop();
		if (x == target)
			break;

		for (int i = 0; i < graph[x].size(); i++) {
			if (!visited[graph[x][i].second]) {
				visited[graph[x][i].second] = true;
				pq.push(graph[x][i]);
			}
		}
	}
}

int main()
{
	int v = 4;
	graph.resize(v);
	
	addedge(0,1,3);
	addedge(0,2,6);
	addedge(1,3,5);
	addedge(2,3,6);

	int source = 0;
	int target = 3;

	best_first_search(source, target, v);

	return 0;
}

File: EXP_5/bfs.py
from queue import PriorityQueue
v = 14
graph = [[] for i in range(v)]


def addEdges():  #node, dest, cost
	addEdgetoGraph(0, 1, 3)
	addEdgetoGraph(0, 2, 6)
	addEdgetoGraph(0, 3, 5)
	addEdgetoGraph(1, 4, 9)
	addEdgetoGraph(1, 5, 8)
	addEdgetoGraph(2, 6, 12)
	addEdgetoGraph(2, 7, 14)
	addEdgetoGraph(3, 8, 7)
	addEdgetoGraph(8, 9, 5)
	addEdgetoGraph(8, 10, 6)
	addEdgetoGraph(9, 11, 1)
	addEdgetoGraph(9, 12, 10)
	addEdgetoGraph(9, 13, 2)


def addEdgetoGraph(x, y, cost):
	graph[x].append((y, cost))
	graph[y].append((x, cost))

def best_first_search(source, target, n):
	visited = [0] * n
	visited[source] = True
	pq = PriorityQueue()
	pq.put((0, source))
	while pq.empty() == False:
		u = pq.get()[1]
		print(u, end=" ")
		if u == target:
			break

		for v, c in graph[u]:
			if (visited[v] == False ):
				visited[v] = True
				pq.put((c, v))
	print()



addEdges()
source = 3 
target = 12
best_first_search(source, target, v)
